site/content/blog/dagger-and-kapp.md

```diff
@@ -0,0 +1,184 @@
+---
+title: Kapp and Dagger"
+slug: dagger-and-cloak
+date: 2022-07-30
+author: Renu Yarday
+excerpt: "In this article, we will explore how to leverage kapp in a Dagger pipeline."
+image: /img/kapp.svg
+tags: ['kapp', 'dagger']
+---
+
+# Running kapp in a Dagger pipeline
+
+In this article we will explore how to leverage kapp in a Dagger pipeline.
+
+### What is Dagger?
+Dagger is a portable devkit to build powerful CI/CD pipelines quickly and run them anywhere. Read more about Dagger [here](https://dagger.io/).
+
+### Introducing kapp package for Dagger
+Do you want to observe your Kubernetes deployments are going smoothly in your Dagger pipeline? Then we highly recommend trying out [kapp deploy](https://github.com/dagger/dagger/tree/main/pkg/universe.dagger.io/alpha/kubernetes/kapp). Kapp is now available as an alpha package with Dagger and can be easily consumed in your CI/CD. 
+
+Below are the steps that one could use to add kapp to a dagger pipeline.
+
+### The micro services demo project
+Leveraging the well known [microservices demo](https://github.com/GoogleCloudPlatform/microservices-demo) to create dagger plan to deploy. The cloned and updated project to run in a local kind cluster is present [here](https://github.com/renuy/microservices-demo).
+
+This project's deployment manifest(configuration) for kubernetes is available as ./release/kubernetes-manifests.yaml
+
+### Build and run locally using dagger plan
+We need to have a dagger plan in place to deploy the application in the cluster. In this blog since the project is already created and build available as image, we will extend the project using the dagger plan to deploy it to a cluster (kind cluster).
+
+#### Deploy.cue
+The dagger plan is written in cue allowing it to be simple and readable. Lets look at the anatomy of the plan.
+
+``` yaml
+import (
+	"dagger.io/dagger"
+    "universe.dagger.io/alpha/kubernetes/kapp"
+)
+
+dagger.#Plan & {
+	actions: test: {
+		deploy: kapp.#Deploy & {
+			app:        "boutique"
+			fs:         client.filesystem."./".read.contents
+			kubeConfig: client.commands.kc.stdout
+			file:       "./release/kubernetes-manifests.yaml"
+		}
+		ls: kapp.#List & {
+			fs:         client.filesystem."./".read.contents
+			kubeConfig: client.commands.kc.stdout
+			namespace:  "default"
+		}
+		inspect: kapp.#Inspect & {
+			app:        "boutique"
+			fs:         client.filesystem."./".read.contents
+			kubeConfig: client.commands.kc.stdout
+		}
+		delete: kapp.#Delete & {
+			app:        "boutique"
+			fs:         client.filesystem."./".read.contents
+			kubeConfig: client.commands.kc.stdout
+		}
+	}
+
+	client: {
+		commands: kc: {
+			name: "kubectl"
+			args: ["config", "view", "--raw"]
+			stdout: dagger.#Secret
+		}
```

Comment:
hmm, is there a better way to obtain kubeconfig? i haven't seen many use `kubectl config view` to grab kubeconfig.

Reply:
Please suggest, a better way. another way of doing could be by passing it as an ENV variable, but I prefer this. others also seem to have adopted this way: e.g. https://github.com/dagger/dagger/blob/e30add5a6e24aa1262b5a2f303dd1929ab3d6597/pkg/universe.dagger.io/x/david%40rawkode.dev/kubernetes/example.cue

Reply:
i guess its fine if that's what other examples are showing.